central/imageintegration/service/service_impl.go

```diff
@@ -332,6 +343,27 @@ func (s *serviceImpl) validateIntegration(ctx context.Context, request *storage.
 }
 
 func (s *serviceImpl) reconcileUpdateImageIntegrationRequest(ctx context.Context, updateRequest *v1.UpdateImageIntegrationRequest) error {
+	var integration *storage.ImageIntegration
+	var exists bool
+	var err error
+
+	// Use the presence of ID to determine if this request is for an updated integration.
+	if updateRequest.GetConfig().GetId() != "" {
```

Comment:
The goal was to only pull the integration from the database once.  Currently there are two repeated DB queries in the update flow for the same integration, the refactor makes it so we don't add a third unnecessarily.

Also unfortunately we cannot rely on the API verbs to determine if a request is an update or new. Which resulted in this added complexity.  

The logic on line 370 still acts as it did before.  If the `updateRequest.GetConfig().GetId() != ""` condition on 351 is `false` (meaning no ID or config) then the conditions on 370 (`updateRequest.GetConfig() == nil`) and/or 373 (`updateRequest.GetConfig().GetId() == ""`) will be `true` and execute as they did before to provide a specific error message to what's missing.

If the `updateRequest.GetConfig().GetId() != ""` condition on 351 is `true` (meaning the config and ID exist) then the lower conditions on lines 370 and 373 will both be `false` and the reconcile executed as today.

Reply:
Hmm, it's unfortunate that we have [the PutImageIntegration API](https://github.com/stackrox/stackrox/blob/master/proto/api/v1/image_integration_service.proto#L53-L59) does not correctly use the `config.id` within it's API path (like [the update integration does](https://github.com/stackrox/stackrox/blob/master/proto/api/v1/image_integration_service.proto#L79)). Looking just at the API definition, I don't think this even works as expected and probably the usage of the PutImageIntegration API is rather limited...

However, I don't think the indirection of `updateRequest.GetConfig().GetId() != ""` (or rather the short-circuit of `updateRequest.GetUpdatePassword() == true` seems right: all other code relies on an ID being set.

So, I'd say let's remove the condition to begin with, since even in the case of `updateRequest.GetUpdatePassword()` the ID _has_ to be set.

Reply:
> So, I'd say let's remove the condition to begin with, since even in the case of `updateRequest.GetUpdatePassword()` the ID has to be set.

If we do this then no integrations can be created without skipping tests - I had originally tried this and discovered the UI calls `TestUpdatedImageIntegration` for new and updated integrations (`updated` in this method name does not indicate the integration was updated, I'm assuming it was named this way to indicate this API is an `updated` version of the old API). So an ID will NOT be set if the integration is a new, and without the `updateRequest.GetConfig().GetId() != ""` check all tests will fail because an integration will not exist.

Unless i'm missing something, to remove the condition would also require a UI and possibly API change to differentiate between an create vs. update request (which is the intent of the ID check).

Reply:
Some recorded API calls + payloads made for each of the image integration actions in the UI, of note is a `updatePassword` is `true` when ID is `""` for new integrations.

#### New Integration - Test
```
POST v1/imageintegrations/test/updated

{"config":{"id":"","skipTestIntegration":false},"updatePassword":true}
```

#### Existing Integration - Test w/ update to stored credentials
```
POST v1/imageintegrations/test/updated

{"config":{"id":"29576d25-9bb2-41c9-a729-79e944511560","skipTestIntegration":false},"updatePassword":true,"id":"29576d25-9bb2-41c9-a729-79e944511560","skipTestIntegration":false}
```

#### Existing Integration - Test w/ NO update to stored credentials
```
POST v1/imageintegrations/test/updated

{"config":{"id":"29576d25-9bb2-41c9-a729-79e944511560","skipTestIntegration":false},"updatePassword":false,"id":"29576d25-9bb2-41c9-a729-79e944511560","skipTestIntegration":false}``
```

#### New Integration - Save
```
POST v1/imageintegrations (id="")

{"id":"","skipTestIntegration":false}
```

#### Existing Integration - Save w/ update to stored credentials
```
PATCH v1/imageintegrations/29576d25-9bb2-41c9-a729-79e944511560

{"config":{"id":"29576d25-9bb2-41c9-a729-79e944511560","skipTestIntegration":false},"updatePassword":true,"id":"29576d25-9bb2-41c9-a729-79e944511560","skipTestIntegration":false}
```

#### Existing Integration Save w/ NO update to stored credentials
```
PATCH v1/imageintegrations/29576d25-9bb2-41c9-a729-79e944511560

{"config":{"id":"29576d25-9bb2-41c9-a729-79e944511560","skipTestIntegration":false},"updatePassword":false,"id":"29576d25-9bb2-41c9-a729-79e944511560","skipTestIntegration":false}
```

Reply:
FYI, created ROX-21601 to capture some of the API debt mentioned.